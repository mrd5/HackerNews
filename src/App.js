import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';


const list = [
  {
    title: 'react',
    url: 'https://facebook.github.io/react/',
    author: 'mathew dorish',
    num_comments: 3,
    points: 4,
    objectID: 0,
  },
  {
    title: 'redux',
    url: 'https://github.com/reactjs/redux',
    author: 'stan richards, mark macdon',
    num_comments: 2,
    points: 4,
    objectID: 1,
  },
];



class Developer{
  constructor(firstname, lastname){
    this.firstname = firstname;
    this.lastname = lastname;
  }

  getName(){
    return this.firstname + " " + this.lastname;
  }
}

const userService = {
  getUserName(user){
    return user.firstname + " " + user.lastname;
  },
}

const key = "name";
const user = {
  [key]: "mathew",//name: mathew
  key: "mathew",  //key: mathew
}

const largeCol = {
  width: '40%',
};

const medCol = {
  width: '30%',
};

const smallCol = {
  width: '10%',
};



const isSearched = searchTerm => item => 
  item.title.toLowerCase().includes(searchTerm.toLowerCase());
//filter function takes a funtion as input

class App extends Component {//functionalities from Component passed over to App
  constructor(props){//initializes internal component state. called only once
    super(props);

    this.state = {
      list,
      searchTerm: '',
    };

    this.onSearchChange = this.onSearchChange.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
  }


  onSearchChange(event){
    this.setState({searchTerm: event.target.value});
  }

  onDismiss(id){
    const isNotId = item => item.objectID !== id;
    const updatedList = this.state.list.filter(isNotId);

    this.setState({list: updatedList});
  }

 //Component is a class
  render() {

    const { searchTerm, list } = this.state;

    return(
      <div className="page">
        <div className="interactions"> 
          <Search 
            value={searchTerm}
            onChange={this.onSearchChange}
          >
            Search:
          </Search>
        </div>
        <Table 
          list={list} //stored in props to pass to other components
          pattern={searchTerm}
          onDismiss={this.onDismiss}
        />
      </div>
    );
  }
}


//finctional state components
const Search = ({value, onChange, children}) =>
  <form>
      {children}
      <input
        type="text"
        value={value}
        onChange={onChange}
      />
    </form>

const Table = ({list, pattern, onDismiss}) =>
  <div className="table">
      {list.filter(isSearched(pattern)).map(item =>
        <div key={item.objectID} className="table-row">

          <span style={ largeCol }>
            <a href={item.url}>{item.title}</a>
          </span>

          <span style={ medCol }> {item.author} </span>
          <span style={ smallCol }> {item.num_comments} </span>
          <span style={ smallCol }> {item.points} </span>

          <span style={ smallCol }>
            <Button
              onClick = {() => onDismiss(item.objectID)}
              className= "button-inline"
            >
              Dismiss
            </Button>
          </span>
        </div>
      )}
    </div>

const Button = ({onClick, className = '', children}) =>
  <button
    onClick = {onClick}
    className = {className}
    type="button"
  >
    {children}
  </button>



export default App;
