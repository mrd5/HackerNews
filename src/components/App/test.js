import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App, { Search, Button, Table } from './index.js';

//App component has two test
describe('App', () => {

	//it function runs a single test. Checks if app runs without crashing
	it('Renders without crashing', () => {
	  const div = document.createElement('div');
	  ReactDOM.render(<App />, div);
	  ReactDOM.unmountComponentAtNode(div);
	});


	test('Has a valid snapshot', () => {
		//.create creates a snapshot of the App component. Renders and stores in DOM
		const component = renderer.create(
			<App />
		);
		let tree = component.toJSON();

		//Checks if new snapshot matches previous
		expect(tree).toMatchSnapshot();
	})

});
