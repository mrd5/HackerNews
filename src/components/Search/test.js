import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App, { Search, Button, Table } from '../App/index.js';



//Search component also has two tests
describe('Search', () => {

	//First test renders component to DOM, verifies there are no errors
	it('Renders without crashing', () => {
		const div = document.createElement('div');
		ReactDOM.render(<Search>Search</Search>, div);
		ReactDOM.unmountComponentAtNode(div);
	});

	//Stores snapshot, compares with previous snapshot
	test('Has a valid snapshot', () => {
		const component = renderer.create(
			<Search>Search</Search>
		);

		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	})
});
