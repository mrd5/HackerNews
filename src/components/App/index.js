import React, { Component } from 'react';
import axios from 'axios'; //used to make asynchronous requests to remote apis
import './index.css';
import { sortBy } from 'lodash';
import {
  DEFAULT_QUERY,
  DEFAULT_HITSPERPAGE,
  PATH_BASE,
  PATH_SEARCH,
  PARAM_SEARCH,
  PARAM_PAGE,
  PARAM_HITSPERPAGE,
} from '../../constants';
import Button from '../Button';
import Search from '../Search';
import Table from '../Table';

const SORTS = {
  NONE: list => list,
  TITLE: list => sortBy(list, 'title'),
  AUTHOR: list => sortBy(list, 'author'),
  COMMENTS: list => sortBy(list, 'num_comments').reverse(),
  POINTS: list => sortBy(list, 'points').reverse(),
};

const withLoading = (Component) => ({ isLoading, ...rest }) =>
  isLoading
    ? <Loading />
    : <Component { ...rest } />

const ButtonWithLoading = withLoading(Button);

const Loading = () =>
  <div>Loading...</div>


const isSearched = searchTerm => item => 
  item.title.toLowerCase().includes(searchTerm.toLowerCase());
//filter function takes a funtion as input

class App extends Component {//functionalities from Component passed over to App
  _isMounted = false;

  constructor(props){//initializes internal component state. called only once
    super(props);

    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      error: null,
      isLoading: false,
      sortKey: 'NONE',
    };

    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
    this.onSort = this.onSort.bind(this);
  }

  needsToSearchTopStories(searchTerm) {
    return (!this.state.results[searchTerm]);
  }

  fetchSearchTopStories(searchTerm, page = 0){
    this.setState({ isLoading: true});

    axios(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HITSPERPAGE}${DEFAULT_HITSPERPAGE}`)
      .then(result => this._isMounted && this.setSearchTopStories(result.data))
      .catch(error => this._isMounted && this.setState({ error }));
  } //Axios uses an HTTP GET request

  setSearchTopStories(result) {
    const { hits, page } = result;
    const { searchKey, results } = this.state;
 
    const oldHits = results && results[searchKey]
      ? results[searchKey].hits
      : [];

    const updatedHits = [
      ...oldHits,
      ...hits
    ];

    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page }
      },
      isLoading: false
    });
  }

  onSearchSubmit(event) {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });

    if (this.needsToSearchTopStories(searchTerm)){
      this.fetchSearchTopStories(searchTerm);
    }

    event.preventDefault();
  }

  onSearchChange(event) {
    this.setState({searchTerm: event.target.value});
  }

  onDismiss(id) {
    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];

    const isNotId = item => item.objectID !== id;
    const updatedHits = hits.filter(isNotId);

    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page }
      }
    });
  }

  onSort(sortKey){
    this.setState({ sortKey });
  }

  componentDidMount() {
    this._isMounted = true;
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    this.fetchSearchTopStories(searchTerm);
  }

  componentWillUnmount() {
    //If user navigates away from page while there is still a pending request
    //But this will prevent that 
    this._isMounted = false;
  }

  //Component is a class
  render() {
    const {
        searchTerm,
        results,
        searchKey,
        error,
        isLoading,
        sortKey,
      } = this.state;

    const page = (
      results && results[searchKey] && results[searchKey].page
    ) || 0;

    const list = (
      results && results[searchKey] && results[searchKey].hits
    ) || [];


    return(
      <div className="page">
        <div className="interactions">
          <Search 
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
          >
            Search:
          </Search>
        </div>
        { error 
          ? <div classname="interactions">
              <p>Something went wrong! Please try again.</p> 
            </div>
          : <Table
            list={ list }
            sortKey={sortKey}
            onSort={this.onSort}
            onDismiss={ this.onDismiss }
          />
        }
        <div className="interactions">
          <ButtonWithLoading
            isLoading={isLoading}
              onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}>
              More
            </ButtonWithLoading>
        </div>
      </div>
    );
  }
}





export default App;

export {
  Button,
  Search,
  Table,
};