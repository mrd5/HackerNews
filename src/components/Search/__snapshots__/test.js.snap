// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Has a valid snapshot 1`] = `
<form
  onSubmit={undefined}
>
  <input
    onChange={undefined}
    type="text"
    value={undefined}
  />
  <button
    type="submit"
  >
    Search
  </button>
</form>
`;
